### ISSUE TITLE
HTTP server incorrectly responds HTTP/1.0 to 1.0 clients

### ISSUE THREAD

**Issue by xkr47** [eclipse/vert.x#1251]
If I send an HTTP/1.0 request to a (Ceylon) Vert.x http server with Vert.x 3.2.0 like

    GET / HTTP/1.0

.. then the server responds with:

    HTTP/1.0 200 OK

However the response version should indicate the capabilities of the server. Since Vert.x supports 1.1 as well it should instead respond with:

    HTTP/1.1 200 OK

This way the client becomes aware that the server also supports HTTP/1.1. It does not however change the semantics of the response - the server still has to send a response that complies with HTTP/1.0 semantics since the client has indicated the maximum version it can use.

The only required change to implement this is thus to always write "HTTP/1.1" in the response. All the logic on setting response headers and closing connections etc should work as before, i.e. use the version from the request to decide semantics.

Some references for more information:
* https://bugs.php.net/bug.php?id=42779
* http://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.1 - "The protocol versioning policy is intended to allow the sender to indicate the format of a message and its capacity for understanding further HTTP communication"

**Comment by testn** [comment]
from RFC2145

```
An HTTP server SHOULD send a response version equal to the highest
   version for which the server is at least conditionally compliant, and
   whose major version is less than or equal to the one received in the
   request.  An HTTP server MUST NOT send a version for which it is not
   at least conditionally compliant.  A server MAY send a 505 (HTTP
   Version Not Supported) response if cannot send a response using the
   major version used in the client's request.

   An HTTP server MAY send a lower response version, if it is known or
   suspected that the client incorrectly implements the HTTP
   specification, but this should not be the default, and this SHOULD
   NOT be done if the request version is HTTP/1.1 or greater.
```

**Comment by xkr47** [comment]
Good, and also:

    One consequence of these rules is that an HTTP/1.1 message sent
    to an HTTP/1.0 recipient (or a recipient whose version is
    unknown) MUST be constructed so that it remains a valid HTTP/1.0
    message when all headers not defined in the HTTP/1.0
    specification [1] are removed.

**Comment by vietj** [comment]
it seems appropriate to allow the override of the HttpServerResponse version and allow the server code to reply a specific version : 1.0 or 1.1 for 1.0 and 1.1 for 1.1 . That would allow the developer to decide the behavior of his server.

**Comment by pmlopes** [comment]
we can fix this at: 

https://github.com/eclipse/vert.x/blob/master/src/main/java/io/vertx/core/http/impl/HttpServerResponseImpl.java#L82

In that line if `version == 1.0 ? 1.1 : version`

However from the RFC this is a should, it is not strictly mandatory to work this way... What would we win if we add that condition?

**Comment by vietj** [comment]
allowing to chose the compatible returned version can be useful if you write an http proxy and you need to return the same code sent by the proxied server

**Comment by purplefox** [comment]
+1 to julien.

Vert.x is a toolkit for building applications it is not designed to be an out of the box fully compliant web server.

Vert.x should let you *build* fully compliant servers, and it should also let you build *faulty* servers (e.g. if you wanted to simulate bad servers for exploit testing), and it should let you build "faulty" clients too.

**Comment by pmlopes** [comment]
IMO it feels a bit a strange if we are going to allow to build *faulty* HTTP servers in regards to this and yet we do not allow any other HTTP methods [RFC2616](https://datatracker.ietf.org/doc/rfc2616/) than the standard 8 plus [RFC5789](https://datatracker.ietf.org/doc/rfc5789/) 1:

* OPTIONS
* GET
* HEAD
* POST
* PUT
* DELETE
* TRACE
* CONNECT
* PATCH

Where we are ignoring [RFC2518](https://datatracker.ietf.org/doc/rfc2518/), [RFC3253](https://datatracker.ietf.org/doc/rfc3253/), [RFC3648](https://datatracker.ietf.org/doc/rfc3648/), [RFC3744](https://datatracker.ietf.org/doc/rfc3744/) and [RFC5323](https://datatracker.ietf.org/doc/rfc5323/) to name a few extensions...

**Comment by vietj** [comment]
I was thinking the same when I saw the message on the group mentioning unknown http methods. Somehow we should have an "UNKNOWN" method and provide an extra rawMethod.

**Comment by xkr47** [comment]
@purplefox I agree with "toolkit for building applications" but I don't think vertx should particularly support faulty implementations. At least personally I will easily revert to use a NetSocket and write http messages by hand if I need to test "broken" behaviour..
